### PR TITLE
experimental SDK support for all our API

### PR DESCRIPTION
**Pull request by toshok**
Particularly interested in api feedback.

This PR adds clients as properties on the libhoney instance.  So:

```javascript
import libhoney from "libhoney";

let honey = new libhoney({...});

honey.events.* // the current event API.  events, builders, etc.
honey.markers.* // CRUD for the markers API
honey.triggers.* // CRUD for the triggers API
honey.boards.* // CRUD for the boards API
```

`honey.events.*` holds all the event API, but the existing methods off libhoney still exist (and call through the events client).  They can either live there forever or we can deprecate them at some later date (and have folks move to using `honey.events`.

Each of the non-events clients is implemented using `ResourceClient` (`api/resource_client.js`), which exposes 4 methods:

1. `create(resource[, datasetName])`
2. `get(id[, datasetName])`
3. `update(id[, datasetName])`
4. `delete(id[, datasetName])`
5. `list([datasetName])`

resources are either dataset scoped (markers, triggers) or not (boards).  For dataset-scoped resources, the `datasetName` must be supplied.  For non-scoped resources, `datasetName` cannot be supplied.  The methods throw if those two statements don't hold.

The methods all return promises.

The types for building each resource are found in `api/${resource}/index.js.`

includes a simple example of listing boards:

```javascript
const Libhoney = require("libhoney");

const honey = new Libhoney({...});

honey.boards
  .list()
  .then(boards =>
    boards.forEach(b => {
      console.log(`Name: ${b.name}`);
      console.log(b.queries);
    })
  )
  .catch(err => {
    console.error(`got error ${err} :( :( :(`);
  });
```

No validation is done client-side.  Server side validation is pretty full featured and better spec'ed, and I don't want to duplicate the validation (in particular I don't want to have to update client side validation if/when server-side changes.)

Also switch all docs/code to prefer `API_KEY`/`apiKey` to `WRITE_KEY`/`writeKey`, falling back to the latter if the former isn't there.

Everything in this PR _should_ be backward compatible (strictly additions), so we don't need to do a major bump.  Tasks to think about when doing a later major bump:

* removing the events methods on the libhoney instance.
* removing all writekey fallbacks.

### CODE REVIEW

File: examples/list-boards/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+const Libhoney = require("libhoney");
+const process = require("process");
+
+const honey = new Libhoney({
+  apiHost: process.env.HONEYCOMB_API_HOST,
+  apiKey: process.env.HONEYCOMB_API_KEY
+});
+
+honey.boards
+  .list()
```
Comment (samstokes) [inline examples/list-boards/index.js]: How do you imagine this evolving to support filtering the list of boards? Would it be `honey.boards.list({filter: 'mine'})` or similar?

In a Ruby API I could imagine it looking something like

```ruby
boards = honeycomb.boards.mine
# as a shortcut for
#      boards = honeycomb.boards.all(filter: 'mine')
boards.each {|board| puts board.name }
```

File: src/api/queries/index.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
+/* notyet
+const validFilterOps = {
```
Comment (samstokes) [inline src/api/queries/index.js]: TODO?

File: src/api/queries/index.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+      this.column = "*";
+    }
+  }
+
+  static fromJSON(c) {
+    return new Calculation(c.column, c.op);
+  }
+}
+export function calculation({ column, op }) {
+  return new Calculation(column, op);
```
Comment (samstokes) [inline src/api/queries/index.js]: This seems like a lot of `export`ed ways to make a calculation - seems like all of the following are user-accessible ways to make a `P95(response_time)`:

```javascript
import { Calculation, calculation } from '@libhoney/api/queries';

calculation.P95('response_time');
calculation({op: 'P95', column: 'response_time'});
new Calculation('response_time', 'P95');
```

Should we bless just one (my vote: `calculation.P95('response_time')`) and not export the others? (Or am I missing something and that is already the case?)

File: src/api/triggers/index.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
+import { Query } from "../queries";
+
+// const validThresholdOps = {
+//   ">": true,
```
Comment (samstokes) [inline src/api/triggers/index.js]: TODO?

**Comment by lielran** [comment]
are you going to continue working on that soon? I'm missing the marker integration and thought of impl this part.

**Comment by toshok** [comment]
closing this mostly so it doesn't continue to be in the list of open PRs.  it's something I want, but we need larger coordination on.  soon, I hope.

**Pull request closed without merging**